ir/src/constraints/mod.rs

```diff
@@ -288,7 +288,9 @@ impl Constraints {
                 // trace segment inference defaults to the lowest segment (the main trace) and is
                 // adjusted according to the use of random values and trace columns.
                 if lhs.trace_segment() < rhs.trace_segment() {
-                    return Err(SemanticError::InvalidUsage("Random values cannot be used in boundary constraints defined against prior trace segments".to_string()));
+                    return Err(
+                        SemanticError::random_value_in_constraint_against_prior_trace_segment(),
+                    );
```

Comment:
I think we should simplify/generalize this error to something like "The constraint expression cannot be enforced against trace segment {}." and maybe call the error `trace_segment_mismatch` and pass in the `lhs.trace_segment()`